src/geode/mesh/core/regular_grid.cpp

```diff
@@ -132,6 +143,73 @@ namespace geode
             return absl::nullopt;
         }
 
+        index_t nb_vertices( index_t direction ) const
+        {
+            return cells_number_.at( direction ) + 1;
+        }
+
+        index_t vertex_index( const Index& index ) const
+        {
+            index_t vertex_id{ 0 };
+            for( const auto d : LRange{ dimension } )
+            {
+                OPENGEODE_ASSERT( index[d] < cells_number_[d] + 1,

```

Comment:
I do not really like the use of cells_number + 1 everywhere. Maybe use your new nb_vertices methods to avoid this implementation detail in the entire class.